### PR TITLE
Update github action for multiple deployment targets

### PR DESCRIPTION
**Pull request by riatzukiza**
### Summary of Changes

Updates the GitHub action for deployment via label, and adds the circle CI jobs necessary to deploy the new applications as laid out in issue #848

### How to Test

1. Add any of the following labels to this PR: 
  + Deploy with CircleCI-raft
  + Deploy with CircleCI-qasp
  + Deploy with CircleCi-a11y
  + Deploy with CircleCi-sandbox
  
2. Which ever label you used, once the pr has passed all its tests, visit the frontend url corresponding (https://tdp-frontend-a11y.app.cloud.gov/ for example, if you picked Deploy with CircleCi-a11y) 

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
  - [x] The following deployment targets are available in the dev space `sandbox`, `raft`, `qasp`, `a11y`
  - [x] Each of those targets are available when deployed at `tdp-[frontend|backend]-${target}.app.cloud.gov` (ie. tdp-frontend-sandbox.cloud.gov)
  - [x] Assigning designated labels for each of those targets triggers a deploy to dev space

As facilitator/product manager, @kniz-raft will decide if ACs are met from Raft's perspective.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

N/A

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

N/A

### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

N/A

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

> NOTE: until we have a proper staging environment this may not be satisfiable prior to merging

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

N/A

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -319,6 +319,49 @@ jobs:
           cf-username: CF_USERNAME_PROD
           frontend-appname: tdp-frontend-prod
 
+  deploy-raft-review:
+    executor: docker-executor
+    working_directory: ~/tdp-deploy
+    steps:
+      - deploy-cloud-dot-gov:
+          backend-appname: tdp-backend-raft-review
+          cf-password: CF_PASSWORD_RAFT_REVIEW
```
Comment (jtwillis92) [inline .circleci/config.yml:328]: The `cf-space`, `cf-username` and `cf-password` can remain as the defaults for all of these new dev environments. We won't need to use separate credentials for apps in the same space.

Comment (riatzukiza) [reply .circleci/config.yml]: Awesome, fixed it

**Review by jtwillis92 — APPROVED** [review]
Tested this successfully for each label target. I did notice a couple caveats however:

* Our space has a route quota of 10, so we don't have enough routes to hold all these apps and the existing `tdp-backend/tdp-frontend` apps at the same time. This will not be an issue once we remove the pre-existing apps after this merges but it should be noted during testing of this PR one or more routes may need to be deleted in order to see all of the targets at the same time.  
* On initial deployments each of these apps will create a new JWT_KEY which must be manually updated in Login.gov, after the initial deployment they will retain the same key. For this round I manually set the JWT_KEY and JWT_CERT equal to what was configured already on `tdp-backend` to avoid needing multiple certificates configured in Login.gov. For posterity, here are the commands I ran:
```
function get_cg_jwt_key () {
    raw=$(cf env tdp-backend | sed -n -e '/JWT_KEY/,/END PRIVATE KEY/ p')
    priv_key=${raw#"JWT_KEY: "}
    echo "$priv_key" | openssl enc -A -base64
}

JWT_KEY=$(get_cg_jwt_key)

raw=$(cf env tdp-backend | sed -n -e '/JWT_CERT/./END CERTIFICATE/ p')
JWT_CERT=${raw#JWT_CERT: "}

cf set-env tdp-backend-a11y JWT_KEY $JWT_KEY
cf set-env tdp-backend-a11y JWT_CERT $JWT_CERT
cf restage tdp-backend-a11y

cf set-env tdp-backend-raft JWT_KEY $JWT_KEY
cf set-env tdp-backend-raft JWT_CERT $JWT_CERT
cf restage tdp-backend-raft

cf set-env tdp-backend-sandbox JWT_KEY $JWT_KEY
cf set-env tdp-backend-sandbox JWT_CERT $JWT_CERT
cf restage tdp-backend-sandbox

cf set-env tdp-backend-qasp JWT_KEY $JWT_KEY
cf set-env tdp-backend-qasp JWT_CERT $JWT_CERT
cf restage tdp-backend-qasp
```
This is fine as long as the apps remain present, but if any are torn down this will need to be done again or we will need to update the JWT_CERT in Login.gov's Dashboard. This is relevant to the stretch goal originally noted in the issue which has been opened as a follow up issue: https://github.com/raft-tech/TANF-app/issues/1060

**Comment by ADPennington** [comment]
very excited about this one 😃

**Comment by ADPennington** [comment]
looks good @riatzukiza! couple of questions for you and @jtwillis92:

- since there are new routes introduced here, can we update the table in the [ADR](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/Architecture%20Decision%20Record/015-development-environments.md) to include these routes?
- Also, the abovementioned ADR indicates reductions in resources that have not yet been implemented. Is the plan to reduce these resources as part of the #1060 work since it will be focused on keeping the dev space tidy?
- how (if at all) will we be using the `deploy with circleci` label going forward? Asking because it's still referenced [here](https://github.com/raft-tech/TANF-app/blob/b7ea650a2e7b879b826934e8790a383cbb24e27e/.github/workflows/deploy-on-label.yml#L32)

**Comment by jtwillis92** [comment]
> * since there are new routes introduced here, can we update the table in the [ADR](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/Architecture%20Decision%20Record/015-development-environments.md) to include these routes?

Will update, thanks for catching the route name change!

> * Also, the abovementioned ADR indicates reductions in resources that have not yet been implemented. Is the plan to reduce these resources as part of the [Automate inactivity teardown for Cloud.gov deployments in the dev space #1060](https://github.com/raft-tech/TANF-app/issues/1060) work since it will be focused on keeping the dev space tidy?

There is a bit of a typo in that ADR that I'll update, but its referring to `memory` allotments not the disk space. Quotas in Cloud.gov are much more restrictive on memory than disk space, we are limited to 3.9GB on the dev space in total. However, we currently use even less memory in the backend than mentioned in the ADR: https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/tdrs-backend/manifest.buildpack.yml#L4 and on the frontend as well https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/tdrs-frontend/manifest.buildpack.yml#L7

> * how (if at all) will we be using the `deploy with circleci` label going forward? Asking because it's still referenced [here](https://github.com/raft-tech/TANF-app/blob/b7ea650a2e7b879b826934e8790a383cbb24e27e/.github/workflows/deploy-on-label.yml#L32)

After this PR is merged we can remove the label that doesn't have the environment at the end. We still need that line to reference it because that filters out all labels that don't start with `Deploy with CircleCI`. The new labels are the only ones that will work:
* Deploy with CircleCI-a11y
* Deploy with CircleCI-raft
* Deploy with CircleCI-sandbox
* Deploy with CircleCI-qasp

The environment name is then split off the end of the label to determine the correct deployment target. If a new label is added that doesn't match our Circle CI jobs configuration then it will fail to run, preventing arbitrary label assignments outside of these.

**Review by ADPennington — APPROVED** [review]
thanks @riatzukiza and @jtwillis92 👍🏾 

@abottoms-coder this is ready to merge!

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
  - [x]  The following deployment targets are available in the dev space `sandbox`, `raft`, `qasp`, `a11y`
  - [x]  Each of those targets are available when deployed at `tdp-[frontend|backend]-${target}.app.cloud.gov` (ie. tdp-frontend-sandbox.cloud.gov)
  - [x]  Assigning designated labels for each of those targets triggers a deploy to dev space
As Product Owner, @lfrohlich will decide if ACs are met.

### Deliverable 2: Tested Code

_n/a: ci changes only_ 

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [x] Does this PR change any linting or CI settings?

### Deliverable 4: Accessible

_n/a_

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

_[ADR #15](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/Architecture%20Decision%20Record/015-development-environments.md) will now include new routes to be used if we deploy using labels_ cc: @ttran-hub

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?
_no new issues detected_

**Pull request merged**